File: Des/c++/decrypt.cpp
#include "header.h"

void decrypt(string*encrypted_file,string*decrypted_file){

    ifstream inp(*encrypted_file,ios_base::binary);
    ofstream out(*decrypted_file,ios_base::binary);

    byte* ip = read_ip(&inp);
    byte* r_ip=reverse_ip(ip,64);
    int64* keys = read_keys(&inp);
    byte** matrix = read_matrix(&inp);

	int file_size;
	inp.read(reinterpret_cast<char*>(&file_size), 4);

	reverse(keys,16);

	cout<<"Count of rounds[1:16] = ";
	int rounds=0;
	cin>>rounds;

	int64 b=0,pi=0;
	for (int i = 0; i < ceil(file_size/8.0); ++i) {
	    inp.read(reinterpret_cast<char*>(&b),8);
	    pi = process_block(b,ip,r_ip,matrix,keys+(16-rounds),rounds);
		out.write(reinterpret_cast<char*>(&pi), i*8 <=(file_size -8)? 8: (file_size+8)%8 );
	}

	inp.close();
	out.close();
	delete [] keys;
	delete [] ip;
	for (int i = 0; i < 4; ++i)
	delete []matrix[i];
	delete [] matrix;
}

File: Des/python/Des/decrypt.py
#!/usr/bin/python3

from functions import *
import struct 

def decrypt(encrypted_file,decrypted_file):
    
    inp = open(encrypted_file,"rb")
    out = open(decrypted_file,"wb")

    ip = read_ip(inp)
    reversed_ip = reverse_ip(ip)
    keys = read_keys(inp)
    matrix = read_matrix(inp)

    file_size = int.from_bytes(inp.read(4),"little")

    rounds = int(input("Count of rounds = "))
    keys = keys[:rounds]
    keys.reverse()
    

    for i in range(file_size//8):
        c = int.from_bytes(inp.read(8),"little")
        p = process_block(c,ip,reversed_ip,matrix,keys)
        out.write(struct.pack("<Q",p))
    tail = file_size % 8 
    if tail > 0:
        c = int.from_bytes(inp.read(8),"little")
        p = process_block(c,ip,reversed_ip,matrix,keys)
        for i in range(tail):
            out.write(struct.pack("B",0b11111111&(p>>i*8)))

if __name__=="__main__":
    decrypt("./encrypted","./decrypted")


